File: 10_falttenLL/10.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.bottom = None

class Solution:
    def merge2Lists(self, head1, head2):
        dummy = Node(-1)
        temp = dummy

        while head1 and head2:
            if head1.data < head2.data:
                temp.bottom = head1
                head1 = head1.bottom
            else:
                temp.bottom = head2
                head2 = head2.bottom
            temp = temp.bottom
            temp.next = None

        if head1:
            temp.bottom = head1
        else:
            temp.bottom = head2

        return dummy.bottom

    def flatten(self, head):
        if not head or not head.next:
            return head

        head.next = self.flatten(head.next)
        head = self.merge2Lists(head, head.next)

        return head

def build_linked_list(arr):
    if not arr:
        return None

    head = Node(arr[0][0])
    current = head
    for i in range(1, len(arr)):
        current.next = Node(arr[i][0])
        current = current.next

    current = head
    for sublist in arr:
        bottom_current = current
        for val in sublist[1:]:
            bottom_current.bottom = Node(val)
            bottom_current = bottom_current.bottom
        current = current.next

    return head

def print_flattened_list(head):
    while head:
        print(head.data, end=" -> ")
        head = head.bottom
    print("None")

if __name__ == "__main__":
    arr = [
        [5, 7, 8, 30],
        [10, 20],
        [19, 22, 50],
        [28, 35, 40, 45]
    ]

    head = build_linked_list(arr)
    sol = Solution()
    flattened_head = sol.flatten(head)
    print("Flattened linked list:")
    print_flattened_list(flattened_head)

File: 10_falttenLL/10.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
    struct Node* bottom;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    newNode->bottom = NULL;
    return newNode;
}

struct Node* merge2Lists(struct Node* head1, struct Node* head2) {
    struct Node dummy;
    struct Node* temp = &dummy;
    dummy.bottom = NULL;

    while (head1 && head2) {
        if (head1->data < head2->data) {
            temp->bottom = head1;
            head1 = head1->bottom;
        } else {
            temp->bottom = head2;
            head2 = head2->bottom;
        }
        temp = temp->bottom;
        temp->next = NULL;
    }

    if (head1)
        temp->bottom = head1;
    else
        temp->bottom = head2;

    return dummy.bottom;
}

struct Node* flatten(struct Node* head) {
    if (!head || !head->next)
        return head;

    head->next = flatten(head->next);
    head = merge2Lists(head, head->next);

    return head;
}

struct Node* buildLinkedList(int arr[][10], int sizes[], int n) {
    if (n == 0)
        return NULL;

    struct Node* head = createNode(arr[0][0]);
    struct Node* current = head;

    for (int i = 1; i < n; i++) {
        current->next = createNode(arr[i][0]);
        current = current->next;
    }

    current = head;
    for (int i = 0; i < n; i++) {
        struct Node* bottomCurrent = current;
        for (int j = 1; j < sizes[i]; j++) {
            bottomCurrent->bottom = createNode(arr[i][j]);
            bottomCurrent = bottomCurrent->bottom;
        }
        current = current->next;
    }

    return head;
}

void printFlattenedList(struct Node* head) {
    while (head) {
        printf("%d -> ", head->data);
        head = head->bottom;
    }
    printf("None\n");
}

int main() {
    int arr[4][10] = {
        {5, 7, 8, 30},
        {10, 20},
        {19, 22, 50},
        {28, 35, 40, 45}
    };
    int sizes[4] = {4, 2, 3, 4};

    struct Node* head = buildLinkedList(arr, sizes, 4);
    struct Node* flattenedHead = flatten(head);
    printf("Flattened linked list:\n");
    printFlattenedList(flattenedHead);

    return 0;
}